### PR TITLE
Prometheus: Add a keyboard shortcut to toggle all exemplars

### PR DESCRIPTION
**Pull request by bobrik**
Sometimes it's hard to see the quantile line behind all the exemplars (far from the worst example):

image

It's next to impossible to find the line you're looking for and it's too many clicks to turn off exemplars.

This commit adds a `d x` keyboard shortcut to toggle exemplar visibility on all prometheus queries. Unlike with legends, the logic is simpler and it does a pure toggle as opposed to "majority toggle" as with legends.

Since exemplars might not be loaded, this also requires refreshing the data. For the same reason, toggling a single panel is not supported, as it will make the panel data out of sync with the rest of the dashboard.

I'm happy to add tests and whatever else is missing if this makes sense and somebody can point me to examples.

### CODE REVIEW

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 1]
```diff
@@ -1094,6 +1094,20 @@ export class DashboardModel implements TimeModel {
     }
   }
 
+  toggleExemplarsForAll() {
+    for (const panel of this.panels) {
+      for (const target of panel.targets) {
+        if (target.datasource.type !== 'prometheus') {
```
Comment (bobrik) [inline public/app/features/dashboard/state/DashboardModel.ts:1100]: Apparently rows are panels too and they have queries, so this fails for them.

**Comment by bobrik** [comment]
Status update: this PR has been ignored.

**Comment by leeoniya** [comment]
thanks for the PR @bobrik, sorry about the delay. this one fell through the cracks. i've asked the metrics team to take a look. cc @gtk-grafana

**Comment by gtk-grafana** [comment]
@bobrik Thanks for taking a look at this! It looks like we've got a lint issue on the frontend which needs to be fixed before merge, but I'm also seeing a bug in which the "x" key toggle which takes people to explore overrides the "d + x" key command added in this PR, so instead of removing the exemplars when typing "d + x", I'm sometimes (about every other time) moved to the explore page, this makes for a pretty frustrating UX.

**Comment by gtk-grafana** [comment]
I'm reaching out to the dashboard squad and I'll let you know what they think, maybe we can use a different set of keys that doesn't conflict with existing commands.

**Comment by ivanortegaalba** [comment]
> @bobrik Thanks for taking a look at this! It looks like we've got a lint issue on the frontend which needs to be fixed before merge, but I'm also seeing a bug in which the "x" key toggle which takes people to explore overrides the "d + x" key command added in this PR, so instead of removing the exemplars when typing "d + x", I'm sometimes (about every other time) moved to the explore page.

From the dashboard's perspective, it is ok to use `x`. I cannot find any other command that makes sense. It is also namespaces by `d` so this can happen with other shortcuts as well. Actually, I don't know why `x` redirects to explore. It should be `g + x`.

**Comment by gtk-grafana** [comment]
Looks like the 'x' shortcut comes from here: https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/utils/getPanelMenu.ts#L154.
I think we can just remove that shortcut line and this PR is good to go?
@bobrik @ivanortegaalba

**Comment by ivanortegaalba** [comment]
> Looks like the 'x' shortcut comes from here: https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/utils/getPanelMenu.ts#L154.
> 
> I think we can just remove that shortcut line and this PR is good to go?
> 
> @bobrik @ivanortegaalba 

I agree. That shortcut should only work when a panel is selected, and it should be combined with `p x`. So we can remove that shortcut and merge this one.

**Comment by bobrik** [comment]
I replaced the `x` shortcut with `p x` for consistency, made the `typecheck` lint happy, and rebased.

I'm not sure what `betterer` is unhappy about here:

```
・ New issue in
"/drone/src/public/app/features/dashboard/state/DashboardModel.ts"!
・
・   /drone/src/public/app/features/dashboard/state/DashboardModel.ts
・ > 1 |
・     | ^ Unexpected any. Specify a different type.
```

There is no `any` on line 1 there.

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 2]
```diff
@@ -1156,6 +1157,21 @@ export class DashboardModel implements TimeModel {
...
   }
 
+  toggleExemplarsForAll() {
+    for (const panel of this.panels) {
+      for (const target of panel.targets) {
+        if (!(target.datasource && target.datasource.type === 'prometheus')) {
+          continue;
+        }
+
+        const promTarget = target as PromQuery;
```
Comment (gtk-grafana) [inline public/app/features/dashboard/state/DashboardModel.ts:1167]: This is the assertion that is failing the betterer check in drone, you can run yarn betterer -u and commit the results to get the tests passing.

File: public/app/core/services/keybindingSrv.ts [hunk 3]
```diff
@@ -259,7 +259,7 @@ export class KeybindingSrv {
 
     // jump to explore if permissions allow
     if (contextSrv.hasAccessToExplore()) {
-      this.bindWithPanelId('x', async (panelId) => {
+      this.bindWithPanelId('p x', async (panelId) => {
```
[the remark this answers is not in the document]
Comment (bobrik) [reply public/app/core/services/keybindingSrv.ts:262]: Here's how it should work:

* `p x` goes to explore (a panel needs to be in focus)
* `d x` toggles exemplars

Are you seeing something different?

File: public/app/features/dashboard/state/DashboardModel.ts (hunk 2 above)
Comment (bobrik) [reply public/app/features/dashboard/state/DashboardModel.ts:1167]: Thanks, drone is happy now.

File: public/app/core/services/keybindingSrv.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply public/app/core/services/keybindingSrv.ts:262]: Nope this is what I'm seeing as well, I'm used to using `g e` for going to global - explore, but `p x` works too

**Review by gtk-grafana — APPROVED** [review]
Looks good, thanks for all your hard work on this!

**Comment by bobrik** [comment]
Thanks @gtk-grafana!

**Pull request merged**